### ISSUE TITLE
tf.keras loss/metric argument/return specification

### ISSUE THREAD

**Issue by YusukeSuzuki** [tensorflow/tensorflow#36298]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.1
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**

This suggestion eases custom loss and custom metrics implementation restrictions. And it removes dirty/tricky codes for text/TensorBoard logging of single-output-multiple-loss-func from user application.

Losses for `Model.compile()` arguments

```python
# current befavior

# model with single output
model.compile(.., loss=func, ..) # OK
model.compile(.., loss=[func], ..) # OK
model.compile(.., loss=[func1, func2], ..) # Error for single output
# model with two outputs
model.compile(.., loss=func, ..) # OK
model.compile(.., loss=[func1_for_out1, func2_for_out2], ..) # OK
model.compile(.., loss=[(func1_for_out1, func2_for_out1), func3_for_out2], ..) # Error
```

```python
# behavior which I sugest

# model with single output
model.compile(.., loss=[func1, func2], ..)
# OK, func1 and func2 are applied to single output with weight [1., 1.]

# model with two outputs
model.compile(.., loss=[(func1_for_out1, func2_for_out1), func3_for_out2], ..)
# OK, func1 and func2 are applied to output1. func3 is applied to output2
# default weight is [(1., 1.), 1.]
# and joint loss is 1. * func1_for_out1(out1) + 1. * func2_for_out1(out1) + 1. * func3_for_out2(out2)
```

Loss function return value

```python
# current befavior
def my_loss_func(..):
  return [loss1, loss2] # Error
  return (loss1, loss2) # Error
  return dict(my_loss_name1=loss1, my_loss_name2=loss2) # Error
```

```python
# behavior which I sugest
# loss function can return list or tuple
def my_loss_func(..):
  return [loss1, loss2]
  return (loss1, loss2)

model.compile(.., loss=[my_loss_func], loss_weights=[1.2])
# joint loss = loss1 * 1.2 / 2 + loss2 * 1.2 / 2
# and loss names on logging are [my_loss_func_1, my_loss_func_2]

# loss function can return dict
def my_loss_func(..):
  return dict(my_loss_name1=loss1, my_loss_name2=loss2) # Error

model.compile(.., loss=[my_loss_func], loss_weights=[1.2])
# joint loss = loss1 * 1.2 / 2 + loss2 * 1.2 / 2
# and loss names on logging are [my_loss_name1, my_loss_name2]
```

Metrics

```python
# current befavior
# Metric function can't return list or tuple or dict in current version.
def my_metrics(..):
  return [metric1, metric2] # Error
  return (metric1, metric2) # Error
  return dict(metric_name_1=metric1, metric_name_2=metric2) # Error
```

```python
# behavior which I sugest
# Metric function can return list or tuple or dict, and metric names are set properly.
def my_metrics(..):
  return [metric1, metric2] # OK, with names [my_metrics_1, my_metrics_2]
  return (metric1, metric2) # OK, with names  [my_metrics_1, my_metrics_2]
  return dict(metric_name_1=metric1, metric_name_2=metric2) # OK, with names  [metric_name_1, metric_name_2]
```

**Will this change the current api? How?**

No. Old codes will run without behavior changes.

**Who will benefit with this feature?**

This change benefits engineer who is using joint loss and want to log losses separately.
And it benefits engineer who want to calculate multiple losses in single function for performance or code simplification.
And it benefits engineer who want to calculate multiple metrics in single function for performance or code simplification.

**Any Other info.**

**Comment by jvishnuvardhan** [comment]
@YusukeSuzuki Sorry for the late response. Are you still interested in contributing? If yes, please feel free to open a PR in  [keras-team/keras repo.](https://github.com/keras-team/keras/issues) repository.

Please note that Keras development moved to keras-team/keras repository to focus on only keras. Thanks!

Idea is good. But, currently each output is associated with a loss function. If you want to use multiple loss functions with single output, then you could write a custom loss function that uses func1 and func2.

Please feel free to open a PR in the keras-team/keras repo. Thanks!
